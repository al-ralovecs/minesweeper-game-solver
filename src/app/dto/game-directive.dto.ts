export class GameDirectiveDto
{
    command: string;
    count: number;

    constructor(directive: string)
    {
        this.parse(directive);
    }

    private parse(directive: string)
    {
        this.command = 'debug';
        this.count = null;

        switch (true) {
            case directive.includes('silent'):
                this.command = 'silent';
                return;
            case directive.includes('debug'):
                this.command = 'debug';
                return;
            case directive.includes('level'):
                this.command = 'level';
                this.count = 1;

                const countStr: string = directive.replace(/level/g,'').trim();

                if (0 === countStr.length) {
                    return;
                }

                const countNum: number = Math.round(Number(countStr));

                if (0 < countNum && 4 >= countNum) {
                    this.count = countNum;
                }
                return;
            case directive.includes('play'):
                this.command = 'play';
                this.count = 1;

                const count: string = directive.replace(/play/g,'').trim();

                if (0 !== count.length) {
                    this.count = Math.round(Number(count));
                }
                return;
            case directive.includes('stop'):
                this.command = 'stop';
                return;
            default:
                //
        }
    }
}
