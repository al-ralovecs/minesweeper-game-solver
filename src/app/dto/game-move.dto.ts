import ActionDto from '../../minesweeper/dto/action.dto';
import { StrategyType } from '../../minesweeper/strategy/abstract-strategy';

export class GameMoveDto {
    constructor(
        public readonly step: number,
        public readonly disposition: string,
        public readonly value?: ActionDto,
    ) {}

    toString(): string {
        return '\n' + this.step + '\n' +
            this.disposition +
            ( typeof this.value === 'undefined'
                    ? '(?, ?, ?)'
                    : `(${this.value.x}, ${this.value.y}, ${StrategyType[this.value.moveMethod]})`
            ) + ' \n';
    }
}
