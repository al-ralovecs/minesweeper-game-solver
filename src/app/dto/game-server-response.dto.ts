import {BoardParserHelper} from "../helper/board-parser.helper";

export enum GameServerResponseType {
    NewGame,
    GotMap,
    TileCleared,
    GotMine,
    Win,
}

export class GameServerResponseDto
{
    type: GameServerResponseType;
    password: string;
    board: number[][];

    public constructor(response: string)
    {
        this.parse(response);
    }

    private parse(response: string): void
    {
        this.board = null;
        this.password = null;

        switch (true) {
            case 'new: OK' === response:
                this.type = GameServerResponseType.NewGame;
                break;
            case response.includes('map:'):
                this.type = GameServerResponseType.GotMap;
                this.board = BoardParserHelper.parse(
                    response
                        .replace(/map:/g,'')
                        .trim(),
                );
                break;
            case 'open: OK' === response:
                this.type = GameServerResponseType.TileCleared;
                break;
            case 'open: You lose' === response:
                this.type = GameServerResponseType.GotMine;
                break;
            case response.includes('open: You win.'):
                this.type = GameServerResponseType.Win;
                this.password = response
                    .replace(/open: You win. The password for this level is:/g, '')
                    .trim();
                break;
        }
    }
}
