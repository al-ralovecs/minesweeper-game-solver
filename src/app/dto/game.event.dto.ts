import { LogPriority } from '../enum/log.priority.enum';

export class GameEventDto {
    public prefix: string = '. Payload: ';

    constructor(public readonly level: LogPriority, public readonly message: string, public readonly payload?: any) {}

    public get toString(): string {
        const date: Date = new Date();

        const datePart =
            date.getFullYear() + '-' +
            (date.getMonth() + 1) + '-' +
            date.getDate() + ' ' +
            date.getHours() + ':' +
            date.getMinutes() + ':' +
            date.getSeconds();

        const priorityPart = ' [' + LogPriority[this.level] + ']';
        const messagePart = ' ' + this.message;
        const payloadPart = this.getPayloadParsed;

        return datePart + priorityPart + messagePart + payloadPart + '\n';
    }

    private get getPayloadParsed(): string {
        if (typeof this.payload === 'string') {
            return this.prefix + this.payload;
        }

        if (typeof this.payload === 'undefined') {
            return '';
        }

        if (null === this.payload) {
            return '';
        }

        try {
            return this.prefix + this.payload.toString();
        } catch (e) {
           //
        }

        return this.prefix + typeof this.payload;
    }
}
