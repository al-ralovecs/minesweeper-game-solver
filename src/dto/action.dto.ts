import LocationDto from './location.dto';
import { StrategyType } from '../strategy/abstract-strategy';

export enum ActionType {
    Clear,
    Flag
}

export default class ActionDto extends LocationDto
{
    public type: ActionType;
    public moveMethod: StrategyType;
    public bigProbability: number;

    private readonly certainty: boolean;

    public constructor(location: LocationDto, type: ActionType, moveMethod: StrategyType, bigProbability: number) {
        super(location.y, location.x);

        this.type = type;
        this.moveMethod = moveMethod;
        this.bigProbability = bigProbability;

        this.certainty = 0.01 >= Math.abs(this.bigProbability - 1);
    }

    public get isCertainty(): boolean
    {
        return this.certainty;
    }

    public get getAction(): ActionType
    {
        return this.type;
    }
}
