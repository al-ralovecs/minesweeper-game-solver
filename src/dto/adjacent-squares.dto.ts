import LocationDto from './location.dto';
import whileLoopAroundTileDo from '../routine/while.loop-around-tile.do';

export default class AdjacentSquaresDto {
    private readonly location: LocationDto;
    private readonly height: number;
    private readonly width: number;
    private readonly size: number;

    public locations: LocationDto[] = [];

    public constructor(location: LocationDto, height: number, width: number, margin: number) {
        this.location = location;
        this.height = height;
        this.width = width;
        this.size = margin;

        this.init();
    }

    private init(): void {
        whileLoopAroundTileDo(
            this.location.y,
            this.location.x,
            this.height,
            this.width,
            (y, x, location) => {
                this.locations.push(location);
            },
            this.size,
        );
    }
}
