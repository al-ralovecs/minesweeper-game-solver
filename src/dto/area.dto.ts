import LocationSetDto from "./location-set.dto";

export default class AreaDto {
    private area: LocationSetDto;
    private readonly readOnlyAre: LocationSetDto;

    public constructor(area: LocationSetDto)
    {
        this.area = area;
    }
}
