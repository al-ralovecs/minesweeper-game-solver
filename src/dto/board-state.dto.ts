import ActionDto from "./action.dto";
import LocationDto from './location.dto';
import LocationSetDto from "./location-set.dto";
import AreaDto from "./area.dto";
import {AdjacentSquaresDto} from "./adjacent-squares.dto";

export default class BoardStateDto
{
    public readonly height: number;
    public readonly width: number;

    public totalFlags: number = 0;
    public totalFlagsConfirmed: number = 0;
    public numOfHidden: number = 0;

    public flagConfirmed: boolean[][];
    public adjFlagsConfirmed: number[][];
    public adjUnrevealed: number[][];
    public revealed: boolean[][];
    public board: number[][];

    public flagOnBoard: boolean[][];
    public adjFlagsOnBoard: number[][];

    public action: ActionDto[][];
    public actionList: ActionDto[];

    public livingWitnesses: LocationSetDto;

    public adjacentLocations1: AdjacentSquaresDto[][];
    public adjacentLocations2: AdjacentSquaresDto[][];

    public unPlayedMoves: number[];

    public constructor(height: number, width: number)
    {
        this.height = height;
        this.width = width;
    }

    public get getTotalUnrevealedCount(): number
    {
        return this.numOfHidden;
    }

    public get getAllLivingWitnesses(): LocationDto[]
    {
        return this.livingWitnesses.data;
    }

    public countAdjacentUnrevealed(location: LocationDto): number
    {
        return this.adjUnrevealed[location.y][location.x];
    }

    public getUnrevealedArea(witnesses: LocationDto[]): AreaDto
    {
        return new AreaDto(this.getUnrevealedSquaresDo(witnesses));
    }

    private getUnrevealedSquaresDo(witnesses: LocationDto[]): LocationSetDto
    {
        let work: LocationSetDto = new LocationSetDto();

        witnesses.forEach(l => {
            this.getAdjacentSquaresIterable(l).forEach(adj => {
                if (this.isUnrevealed(adj)) {
                    work.add(adj);
                }
            })
        });

        return work;
    }

    private isUnrevealed(location: LocationDto): boolean
    {
        return ! this.flagConfirmed[location.y][location.x] && ! this.revealed[location.y][location.x];
    }

    private getAdjacentSquaresIterable(location: LocationDto): LocationDto[]
    {
        if (undefined === this.adjacentLocations1[location.y][location.x]) {
            this.adjacentLocations1[location.y][location.x] = new AdjacentSquaresDto(
                location,
                this.height,
                this.width,
                1
            );
        }

        return this.adjacentLocations1[location.y][location.x].locations;
    }
}
