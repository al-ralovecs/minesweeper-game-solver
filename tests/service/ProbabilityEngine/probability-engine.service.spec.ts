import { disposition as d1 } from '../../__fixtures__/state/16x16x40.pe.best-probability.json';
import { disposition as d2} from '../../__fixtures__/state/16x16x40.pe.most-info.json';

import Binomial from '../../../src/utility/binomial';
import WitnessWebService from '../../../src/service/witness-web.service';
import DeadLocationsService from '../../../src/service/dead-locations.service';
import ProbabilityEngineService from '../../../src/service/probability-engine.service';

import { ActionType } from '../../../src/dto/action.dto';

import getBoardStateService from '../composite/function.get-board-state.service';
import getWitnessWebService from '../composite/function.get-witness-web.service';
import applyTrivialSearchStrategy from '../composite/function.apply-trivial-search.strategy';
import applyLocalSearchStrategy from '../composite/function.apply-local-search.strategy';
import getDeadLocationsService from '../composite/function.get-dead-locations.service';
import getProbabilityEngineService from '../composite/function.get-probability-engine.service';

describe('ProbabilityEngineService', () => {
    test('process on adopted disposition #1', () => {
        const probabilityEngineService: ProbabilityEngineService = getService(d1);

        expect(probabilityEngineService.getOffEdgeProb).toBe(0.87487);
        expect(probabilityEngineService.cutOffProbability).toBe(0.88858464);
        expect(probabilityEngineService.bestProbability).toBe(0.925609);
    });
    test('process on adopted disposition #1', () => {
        const probabilityEngineService: ProbabilityEngineService = getService(d2);

        expect(probabilityEngineService.getOffEdgeProb).toBe(0.846856);
        expect(probabilityEngineService.cutOffProbability).toBe(0.8726601599999999);
        expect(probabilityEngineService.bestProbability).toBe(0.909021);
    });
});

function getService(disposition: number[][]): ProbabilityEngineService
{
    const mines: number = 40;

    const boardStateService = getBoardStateService(disposition, mines);
    const binomialEngine: Binomial = new Binomial(1000000, 100);
    const witnessWebService: WitnessWebService = getWitnessWebService(boardStateService, binomialEngine);

    applyTrivialSearchStrategy(boardStateService, witnessWebService);

    expect(
        boardStateService.getBoardState.getActions.filter(a => ActionType.Clear === a.type).length
    ).toBe(0);

    applyLocalSearchStrategy(boardStateService, witnessWebService);

    expect(
        boardStateService.getBoardState.getActions.filter(a => ActionType.Clear === a.type).length
    ).toBe(0);

    const deadLocationsService: DeadLocationsService = getDeadLocationsService(
        boardStateService,
        witnessWebService
    );
    
    return getProbabilityEngineService(
        boardStateService, 
        witnessWebService, 
        binomialEngine, 
        deadLocationsService
    );
}
