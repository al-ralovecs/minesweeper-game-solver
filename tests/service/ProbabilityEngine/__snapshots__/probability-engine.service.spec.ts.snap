// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProbabilityEngineService process on adopted disposition #1 1`] = `
Array [
  0.517883,
  0.48211800000000005,
  0.925609,
  0.925609,
  0.666667,
  0.40772600000000003,
  0.925609,
  0.37053,
  0.258942,
  0.37053,
  0.37053,
  0.925609,
  0.543157,
]
`;
